File: reference/src/metrics/store.ts
/**
 * In-memory metric store with per-label tracking.
 */

import type {
  MetricDataPoint,
  MetricDefinition,
  MetricQuery,
  MetricStore,
  MetricSummary,
} from './types.js';

function labelsMatch(
  pointLabels: Record<string, string> | undefined,
  queryLabels: Record<string, string> | undefined,
): boolean {
  if (!queryLabels || Object.keys(queryLabels).length === 0) return true;
  if (!pointLabels) return false;
  return Object.entries(queryLabels).every(([k, v]) => pointLabels[k] === v);
}

export function createMetricStore(): MetricStore {
  const definitions = new Map<string, MetricDefinition>();
  const data = new Map<string, MetricDataPoint[]>();

  function getPoints(metric: string): MetricDataPoint[] {
    let points = data.get(metric);
    if (!points) {
      points = [];
      data.set(metric, points);
    }
    return points;
  }

  return {
    register(definition: MetricDefinition): void {
      definitions.set(definition.name, definition);
    },

    record(partial: Omit<MetricDataPoint, 'timestamp'> & { timestamp?: string }): MetricDataPoint {
      const point: MetricDataPoint = {
        metric: partial.metric,
        value: partial.value,
        timestamp: partial.timestamp ?? new Date().toISOString(),
        labels: partial.labels,
      };
      getPoints(point.metric).push(point);
      return point;
    },

    current(metric: string, labels?: Record<string, string>): number | undefined {
      const points = data.get(metric);
      if (!points || points.length === 0) return undefined;
      for (let i = points.length - 1; i >= 0; i--) {
        if (labelsMatch(points[i].labels, labels)) {
          return points[i].value;
        }
      }
      return undefined;
    },

    query(query: MetricQuery): readonly MetricDataPoint[] {
      const points = data.get(query.metric);
      if (!points) return [];
      return points.filter((p) => {
        if (!labelsMatch(p.labels, query.labels)) return false;
        if (query.from && p.timestamp < query.from) return false;
        if (query.to && p.timestamp > query.to) return false;
        return true;
      });
    },

    summarize(metric: string, labels?: Record<string, string>): MetricSummary | undefined {
      const points = data.get(metric);
      if (!points || points.length === 0) return undefined;
      const matching = points.filter((p) => labelsMatch(p.labels, labels));
      if (matching.length === 0) return undefined;

      const values = matching.map((p) => p.value);
      return {
        metric,
        count: values.length,
        min: Math.min(...values),
        max: Math.max(...values),
        avg: values.reduce((sum, v) => sum + v, 0) / values.length,
        latest: values[values.length - 1],
      };
    },

    snapshot(labels?: Record<string, string>): Record<string, number> {
      const result: Record<string, number> = {};
      for (const [metric, points] of data) {
        for (let i = points.length - 1; i >= 0; i--) {
          if (labelsMatch(points[i].labels, labels)) {
            result[metric] = points[i].value;
            break;
          }
        }
      }
      return result;
    },

    definitions(): readonly MetricDefinition[] {
      return Array.from(definitions.values());
    },
  };
}

File: sdk-python/src/ai_sdlc/metrics/store.py
"""In-memory metric store with per-label tracking."""

from __future__ import annotations

from datetime import UTC, datetime

from .types import MetricDataPoint, MetricDefinition, MetricQuery, MetricSummary


def _labels_match(
    point_labels: dict[str, str] | None,
    query_labels: dict[str, str] | None,
) -> bool:
    if not query_labels:
        return True
    if not point_labels:
        return False
    return all(point_labels.get(k) == v for k, v in query_labels.items())


class _MetricStoreImpl:
    def __init__(self) -> None:
        self._definitions: dict[str, MetricDefinition] = {}
        self._data: dict[str, list[MetricDataPoint]] = {}

    def register(self, definition: MetricDefinition) -> None:
        self._definitions[definition.name] = definition

    def record(
        self,
        metric: str,
        value: float,
        labels: dict[str, str] | None = None,
        timestamp: str | None = None,
    ) -> MetricDataPoint:
        point = MetricDataPoint(
            metric=metric,
            value=value,
            timestamp=timestamp or datetime.now(UTC).isoformat(),
            labels=labels,
        )
        self._data.setdefault(metric, []).append(point)
        return point

    def current(self, metric: str, labels: dict[str, str] | None = None) -> float | None:
        points = self._data.get(metric)
        if not points:
            return None
        for p in reversed(points):
            if _labels_match(p.labels, labels):
                return p.value
        return None

    def query(self, query: MetricQuery) -> list[MetricDataPoint]:
        points = self._data.get(query.metric)
        if not points:
            return []
        result: list[MetricDataPoint] = []
        for p in points:
            if not _labels_match(p.labels, query.labels):
                continue
            if query.from_ and p.timestamp < query.from_:
                continue
            if query.to and p.timestamp > query.to:
                continue
            result.append(p)
        return result

    def summarize(
        self, metric: str, labels: dict[str, str] | None = None
    ) -> MetricSummary | None:
        points = self._data.get(metric)
        if not points:
            return None
        matching = [p for p in points if _labels_match(p.labels, labels)]
        if not matching:
            return None
        values = [p.value for p in matching]
        return MetricSummary(
            metric=metric,
            count=len(values),
            min=min(values),
            max=max(values),
            avg=sum(values) / len(values),
            latest=values[-1],
        )

    def snapshot(self, labels: dict[str, str] | None = None) -> dict[str, float]:
        result: dict[str, float] = {}
        for metric, points in self._data.items():
            for p in reversed(points):
                if _labels_match(p.labels, labels):
                    result[metric] = p.value
                    break
        return result

    def definitions(self) -> list[MetricDefinition]:
        return list(self._definitions.values())


def create_metric_store() -> _MetricStoreImpl:
    """Create an in-memory metric store."""
    return _MetricStoreImpl()
